rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1474,43 +1591,57 @@ extern "C"
     {
-        std::mutex conditionMutex;
-        std::condition_variable conditionVariable;
+        if (!waitset) {
+            return RMW_RET_ERROR;
+        }
+        CustomWaitsetInfo * waitset_info = static_cast<CustomWaitsetInfo*>(waitset->data);
+        if (!waitset_info) {
+            return RMW_RET_ERROR;
+        }
+        std::mutex * conditionMutex = &waitset_info->condition_mutex;
+        std::condition_variable * conditionVariable = &waitset_info->condition;
+        if (!conditionMutex || !conditionVariable) {
+            return RMW_RET_ERROR;
+        }
 
         for(unsigned long i = 0; i < subscriptions->subscriber_count; ++i)
         {
             void *data = subscriptions->subscribers[i];
             CustomSubscriberInfo *custom_subscriber_info = (CustomSubscriberInfo*)data;
-            custom_subscriber_info->listener_->attachCondition(&conditionMutex, &conditionVariable);
+            custom_subscriber_info->listener_->attachCondition(conditionMutex, conditionVariable);
         }
 
         for(unsigned long i = 0; i < clients->client_count; ++i)
         {
             void *data = clients->clients[i];
             CustomClientInfo *custom_client_info = (CustomClientInfo*)data;
-            custom_client_info->listener_->attachCondition(&conditionMutex, &conditionVariable);
+            custom_client_info->listener_->attachCondition(conditionMutex, conditionVariable);
         }
 
         for(unsigned long i = 0; i < services->service_count; ++i)
         {
             void *data = services->services[i];
             CustomServiceInfo *custom_service_info = (CustomServiceInfo*)data;
-            custom_service_info->listener_->attachCondition(&conditionMutex, &conditionVariable);
+            custom_service_info->listener_->attachCondition(conditionMutex, conditionVariable);
         }
 
-        for(unsigned long i = 0; i < guard_conditions->guard_condition_count; ++i)
-        {
-            void *data = guard_conditions->guard_conditions[i];
-            GuardCondition *guard_condition = (GuardCondition*)data;
-            guard_condition->attachCondition(&conditionMutex, &conditionVariable);
+        if (guard_conditions) {
+            for(unsigned long i = 0; i < guard_conditions->guard_condition_count; ++i)
+            {
+                void *data = guard_conditions->guard_conditions[i];
+                GuardCondition *guard_condition = (GuardCondition*)data;
+                guard_condition->attachCondition(conditionMutex, conditionVariable);
+            }
         }
+        rmw_guard_conditions_t * fixed_guard_conditions = waitset->fixed_guard_conditions;
 
-        std::unique_lock<std::mutex> lock(conditionMutex);
+        std::unique_lock<std::mutex> lock(*conditionMutex);
 
         // First check variables.
-        bool hasToWait = true;
+        bool hasToWait = (wait_timeout && wait_timeout->sec > 0 && wait_timeout->nsec > 0) || !wait_timeout;
```

Comment:
The logic seems wrong. A timeout with `sec = 0` but `nsec > 0` would result in `false`?